### ISSUE TITLE
chunked encoding parser is a little bit confusing

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5665]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)| @zooko reported|
|-|-|
|Trac ID|trac#5665|
|Type|enhancement|
|Created|2012-05-03 07:06:21Z|
|Branch|https://github.com/twisted/twisted/tree/chunk-decoder-5665|

I'm hot on the trail of a performance problem in chunked encoding parsing. I have a benchmark and everything! But this ticket is not about that. This ticket is about the fact that, in trying to fix the performance problem in the parser, I was briefly confused by the state of the parser being encoded into two member variables -- ```self.state``` and ```self.finish```. I think representing the five possible states as five possible values of ```self.state``` is easier to understand than as four possible values of ```self.state``` plus one boolean in ```self.finish``` which is used only when ```self.state == 'trailer'```. This is especially true since the condition of ```self.state == 'trailer' and self.finish == False``` doesn't really have anything to do with the chunk encoding trailer -- it is usually just that you're moving on to parsing the next chunk.

This patch makes it clearer to me. It doesn't break any unit tests.

Attachments:

* [more-explicit-states.unidiff.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/527/5278fbbe5c7e115bb00a4102306a54c7cd830404/7c0b9ac47a35950659a9713ef3f85ca790867ff0.txt) (3101 bytes) - added by zooko on 2012-05-03 07:07:59Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5665 5665
type__enhancement enhancement
reporter__zooko zooko
priority__normal normal
milestone__ 
branch__branches_chunk_decoder_5665 branches/chunk-decoder-5665
branch_author__zooko__therve zooko, therve
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1336028781000000 1336028781000000
changetime__1341745811000000 1341745811000000
version__None None
owner__therve therve
cc__jknight cc__zooko@... cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

I applied the patch, and then went a bit further away by splitting dataReceived (and removing the finish flag). It requires another look, though.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra set owner to @therve|
|-|-|

1.Minor thing but you reference `C{'CHUNK-LENGTH'}` where it should be `C{'CHUNK_LENGTH'}. 
 1. it needs a (.bugfix) news file
 1. could you also run it on the buildslaves and post the link here

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve removed owner|
|-|-|

Thanks, both changes made, results here: buildbot.twistedmatrix.com/boxes-supported?branch=/branches/chunk-decoder-5665

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess set owner to @therve|
|-|-|

Thanks for the patch, zooko, and for your additional improvements, therve. I agree that this makes the state machine much clearer.

The news fragment file should either be an empty .misc or a non-empty .bugfix. (See ReviewProcess for details)

Other than that, looks good, please merge!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set status to `closed`|
|-|-|

(In [34696]) Merge chunk-decoder-5665

Authors: zooko, therve
Reviewers: jesstess, thijs
Fixes: [#5665](https://github.com/twisted/twisted/issues/5665)

Make twisted.web.http._ChunkedTransferDecoder.dataReceived a state machine,
introducing a new state to clarify it and remove an additional boolean state.
